File: org.kevoree.modeling.microframework/src/main/java/org/kevoree/modeling/api/json/JSONModelSerializer.java
package org.kevoree.modeling.api.json;

import org.kevoree.modeling.api.Callback;
import org.kevoree.modeling.api.KObject;
import org.kevoree.modeling.api.ModelSerializer;
import org.kevoree.modeling.api.ModelVisitor;
import org.kevoree.modeling.api.data.KStore;

import java.io.*;
import java.util.Set;

/**
 * Created with IntelliJ IDEA.
 * User: duke
 * Date: 28/08/13
 * Time: 11:08
 */

public class JSONModelSerializer implements ModelSerializer {

    public static final String keyMeta = "@meta";

    public static final String keyKid = "@uuid";

    public static final String keyRoot = "@root";

    @Override
    public void serialize(KObject model, final Callback<String> callback) {
        final ByteArrayOutputStream outstream = new ByteArrayOutputStream();
        serializeToStream(model, outstream, new Callback<Throwable>() {
            @Override
            public void on(Throwable e) {
                try {
                    outstream.close();
                    if (e == null) {
                        callback.on(outstream.toString());
                    } else {
                        callback.on(null);
                    }
                } catch (IOException e2) {
                    callback.on(null);
                }
            }
        });
    }

    @Override
    public void serializeToStream(KObject model, OutputStream raw, Callback<Throwable> error) {
        final PrintStream out = new PrintStream(new BufferedOutputStream(raw), false);
        out.print("[\n");
        printJSON(model, out);
        model.graphVisit(new ModelVisitor() {
            @Override
            public VisitResult visit(KObject elem) {
                out.print(",");
                printJSON(elem, out);
                return ModelVisitor.VisitResult.CONTINUE;
            }
        }, new Callback<Throwable>() {
            @Override
            public void on(Throwable throwable) {
                out.print("]\n");
                out.flush();
                error.on(null);
            }
        });
    }


    public static void printJSON(KObject elem, PrintStream builder) {
        builder.append("{\n");
        builder.append("\t\"" + keyMeta + "\" : \"");
        builder.append(elem.metaClass().metaName());
        builder.append("\",\n");
        builder.append("\t\"" + keyKid + "\" : \"");
        builder.append(elem.uuid() + "");
        if (elem.isRoot()) {
            builder.append("\",\n");
            builder.append("\t\"" + keyRoot + "\" : \"");
            builder.append("true");
        }
        builder.append("\",\n");
        for (int i = 0; i < elem.metaAttributes().length; i++) {
            Object payload = elem.get(elem.metaAttributes()[i]);
            if (payload != null) {
                builder.append("\t");
                builder.append("\"");
                builder.append(elem.metaAttributes()[i].metaName());
                builder.append("\" : \"");
                builder.append(payload.toString());
                builder.append("\",\n");
            }
        }
        for (int i = 0; i < elem.metaReferences().length; i++) {
            Object[] raw = elem.view().dimension().universe().storage().raw(elem, KStore.AccessMode.READ);
            Object payload = null;
            if (raw != null) {
                payload = raw[elem.metaReferences()[i].index()];
            }
            if (payload != null) {
                builder.append("\t");
                builder.append("\"");
                builder.append(elem.metaReferences()[i].metaName());
                builder.append("\" :");
                if (elem.metaReferences()[i].single()) {
                    builder.append("\"");
                    builder.append(payload.toString());
                    builder.append("\"");
                } else {
                    Set<Long> elems = (Set<Long>) payload;
                    Long[] elemsArr = elems.toArray(new Long[elems.size()]);
                    boolean isFirst = true;
                    builder.append(" [");
                    for (int j = 0; j < elemsArr.length; j++) {
                        if (!isFirst) {
                            builder.append(",");
                        }
                        builder.append("\"");
                        builder.append(elemsArr[j] + "");
                        builder.append("\"");
                        isFirst = false;
                    }
                    builder.append("]");
                }
                builder.append(",\n");
            }
        }
        // int lastcomma = builder.lastIndexOf(",");
        //  builder.setCharAt(lastcomma, ' ');
        builder.append("}\n");
    }

}

File: org.kevoree.modeling.microframework/src/main/java/org/kevoree/modeling/api/json/JSONModelSerializer.kt
package org.kevoree.modeling.api.json

import org.kevoree.modeling.api.KObject
import java.io.PrintStream
import org.kevoree.modeling.api.ModelVisitor
import org.kevoree.modeling.api.ModelAttributeVisitor
import java.io.OutputStream
import org.kevoree.modeling.api.ModelSerializer
import java.io.ByteArrayOutputStream
import org.kevoree.modeling.api.util.AttConverter
import org.kevoree.modeling.api.Callback

/**
 * Created with IntelliJ IDEA.
 * User: duke
 * Date: 28/08/13
 * Time: 11:08
 */


class ModelReferenceVisitor(val out: PrintStream) : ModelVisitor() {
    override fun beginVisitRef(refName: String, refType: String): Boolean {
        out.print(",\"" + refName + "\":[")
        isFirst = true
        return true
    }
    override fun endVisitRef(refName: String) {
        out.print("]")
    }
    var isFirst = true
    public override fun visit(elem: KObject, refNameInParent: String, parent: KObject) {
        if (!isFirst) {
            out.print(",")
        } else {
            isFirst = false
        }
        out.print("\"" + elem.path() + "\"")
    }
}

public class JSONModelSerializer() : ModelSerializer {

    override fun serialize(model: KObject<Any?>, callback: Callback<String>, error: Callback<Exception>) {
        val outstream = ByteArrayOutputStream()
        serializeToStream(model, outstream, object : Callback<Boolean> {
            override fun on(p: Boolean) {
                outstream.close()
                if (p) {
                    callback.on(outstream.toString())
                } else {
                    error.on(Exception("Unknow Error while during model serialization !"));
                }
            }
        }, error);
    }

    override fun serializeToStream(model: KObject<Any?>, raw: OutputStream, callback: Callback<Boolean>, error: Callback<Exception>) {
        val out = PrintStream(java.io.BufferedOutputStream(raw), false)
        //visitor for printing reference
        val internalReferenceVisitor = ModelReferenceVisitor(out)
        //Visitor for Model navigation
        var masterVisitor = object : ModelVisitor() {
            var isFirstInRef = true
            override public fun beginVisitElem(elem: KObject<*>) {
                if (!isFirstInRef) {
                    out.print(",")
                    isFirstInRef = false
                }
                printAttName(elem, out)
                internalReferenceVisitor.alreadyVisited?.clear()
                elem.visitNotContained(internalReferenceVisitor)
            }
            override public fun endVisitElem(elem: KObject<*>) {
                out.println("}")
                isFirstInRef = false
            }
            override fun beginVisitRef(refName: String, refType: String): Boolean {
                out.print(",\"" + refName + "\":[")
                isFirstInRef = true
                return true
            }
            override fun endVisitRef(refName: String) {
                out.print("]")
                isFirstInRef = false
            }
            public override fun visit(elem: KObject<*>, refNameInParent: String, parent: KObject<*>) {

            }
        }
        model.deepVisitContained(masterVisitor);
        out.flush();
    }

    private fun printAttName(elem: KObject<*>, out: PrintStream) {
        var isRoot = ""
        if (elem.path().equals("/")) {
            isRoot = "root:"
        }
        out.print("\n{\"class\":\"" + isRoot + elem.metaClassName() + "@" + elem.key() + "\"")
        val attributeVisitor = object : ModelAttributeVisitor {
            public override fun visit(value: Any?, name: String, parent: KObject<*>) {
                if (value != null) {
                    out.print(",\"" + name + "\":\"")
                    if (value is java.util.Date) {
                        JSONString.encode(out, "" + value.getTime())
                    } else {
                        JSONString.encode(out, AttConverter.convFlatAtt(value))
                    }
                    out.print("\"")
                }
            }
        }
        elem.visitAttributes(attributeVisitor)
    }

}

